### ISSUE TITLE
Performance issue with OpenSSL 3 and libsrtp

### ISSUE THREAD

**Issue by murillo128** [openssl/openssl#20625]
I am building libsrtp using nodejs bundled OpenSSL and I have seen a performance performance issue with node 18 & 19 which bundles OpenSSL 3.0 both on AES GCM and AES CTR modes compared to node 16 which bundles OpenSSL 1.1 using the following benchmark:

https://gist.github.com/Sean-Der/7a42bd70edfe1324ccc6ab399d653c0e

The performance hit is quite significative, with AES GCM being more than 2x slower and AES CTR being 1.5x slower (depending on the  server types I have tested with):

Server 1:
```
node v16.19.0 - openssl 1.1.1s+quic
gcm test done in 847746ms
ctr test done in 2209247ms

node v18.15.0 - openssl 3.0.8+quic
gcm test done in 2042822
ctr test done in 2837856
```

Server 2:
```
node v16.19.0  - openssl 1.1.1s+quic
gcm test done in 2210976
ctr test done in 3035863

node v18.15.0  - openssl 3.0.8+quic
gcm test done in 4259098
ctr test done in 4081153
```

I think that the openssl usage in libsrtp should not be causing this performance issues:

https://github.com/cisco/libsrtp/blob/main/crypto/cipher/aes_gcm_ossl.c

**Comment by t8m** [comment]
How much data is the test pushing through the encryption context per single init? If the amount of encrypted data per single context init is small the implicit fetch done by using EVP_aes_XXX_gcm() placeholder EVP_CIPHERs will significantly affect the performance. A fetch could be done explicitly using the EVP_CIPHER_fetch() function during initialization of the test to avoid that performance penalty.

**Comment by murillo128** [comment]
@pabuhler pointed out the size of the encrypted test too, so I have made it variable with different payload sizes (+ rtp header):

```
Node 16
0    payload size gcm done in 886714
0    payload size ctr done in 2421151
400  payload size gcm done in 1494584
400  payload size ctr done in 4381984
1200 payload size gcm done in 2326866
1200 payload size ctr done in 8568198

Node 18
0    payload size gcm done in 2007877
0    payload size ctr done in 3287084
400  payload size gcm done in 2705423
400  payload size ctr done in 5485942
1200 payload size gcm done in 3549461
1200 payload size ctr done in 9669171
```
I don't have enough openssl API knowledge to do the changes myself, would it be possible for you point me what should be change in libsrtp file?

**Comment by t8m** [comment]
It really depends on how the libsrtp API is organized, I do not have any deeper knowledge of it. Basically you need to replace the EVP_aes_256_gcm() and EVP_aes_128_gcm() calls with objects that will be pre-fetched (outside of the benchmarking loop) via `EVP_CIPHER_fetch(libctx (can be NULL), "AES-256-GCM", NULL)` and `EVP_CIPHER_fetch(libctx (can be NULL), "AES-128-GCM", NULL)` calls. The returned value is refcounted, so it needs to be freed after it is no longer being used by EVP_CIPHER_free() - i.e., at the termination of the application.

**Comment by murillo128** [comment]
the code for gcm seems to be a simple wrapper to support multiple backends (boringssl,..)

https://github.com/cisco/libsrtp/blob/main/crypto/cipher/aes_gcm_ossl.c

Note that I can't prefecth outside of the testing loop as I need to bencharch the total cpu used to encrypt multiple rtp packets (which is the normal libsrtp usage)

**Comment by t8m** [comment]
I know it would not make sense to prefetch in the benchmark itself, the change would have to be done in libsrtp to do the prefetch (possibly somewhere inside the srtp_create() call).

**Comment by pabuhler** [comment]
I guess it is possible to pre fetch in srtp_aes_gcm_openssl_alloc, in which case it should only be called once in that test, inside the srtp_create() function. @murillo128  will try this out?

**Comment by murillo128** [comment]
@pabuhler If you send me the code change I can definitively test it.

**Comment by murillo128** [comment]
Not seeing much improvements between the main branch and the https://github.com/pabuhler/libsrtp/tree/openssl-3-performence one

```
main branch
   0 payload size gcm done in 2040600
   0 payload size ctr done in 3301378
 400 payload size gcm done in 2690191
 400 payload size ctr done in 5314790
1200 payload size gcm done in 3578534
1200 payload size ctr done in 9631898

openssl-3-performance
   0 payload size gcm done in 1957482
   0 payload size ctr done in 3276440
 400 payload size gcm done in 2611356
 400 payload size ctr done in 5423681
1200 payload size gcm done in 3349710
1200 payload size ctr done in 9591493
```

**Comment by pabuhler** [comment]
@murillo128  so I guess we need to start profiling unless there are some other obvious things we are doing wrong. Are you interested or would you like me to look more into this?

**Comment by murillo128** [comment]
I am definetively interested, but not sure if I have enough knowledge to do it.. :)

it would be great if we could do it together

**Comment by murillo128** [comment]
I will see if I can get a better profiling, but on my server wide load tests I see quite a lot cpu usage in `__stpncpy_sse2_unaligned` `____wcstof_l_internal` and `_aesni_ctr32_ghash_6x` functions that do not happen when using OpenSSL1.1

**Comment by murillo128** [comment]
I have made some test with `perf` and `flamegraphs` and here are the results:

```
node 18 optimized oppenssl3 branch
   0 payload size gcm done in 1979180
   0 payload size ctr done in 3335598
 400 payload size gcm done in 2563691
 400 payload size ctr done in 5442702
1200 payload size gcm done in 3367468
1200 payload size ctr done in 9583309
```
![libsrtp-openssl3](https://user-images.githubusercontent.com/1070835/228836457-d22f8d32-1806-4794-a5f8-308142355b65.svg)

```
node 16 openssl 1.1
   0 payload size gcm done in 896950
   0 payload size ctr done in 2412647
 400 payload size gcm done in 1484701
 400 payload size ctr done in 4492788
1200 payload size gcm done in 2347741
1200 payload size ctr done in 8839961
```
![libsrtp-openssl1 1](https://user-images.githubusercontent.com/1070835/228836460-7f6ba599-d1ee-4a31-9696-c0ddcafc0467.svg)

**Comment by pabuhler** [comment]
this is a simple openssl program that emulates the usages inside of libsrtp,  https://gist.github.com/pabuhler/7beef08da203018ae905944c44c911dd

**Comment by pabuhler** [comment]
@murillo128  I guess so,  will try with 3.1 a bit later today

**Comment by pabuhler** [comment]
Updated the gist, in libsrtp we only set the key and init the ctx with the cipher once, the IV is then updated each time in a loop. This means that pre fetching does not make much difference.
I see the same ~2.5 X slow down with 3.0 and 3.1 branch of openssl.

**Comment by t8m** [comment]
I can reproduce this with the ossl_speed.c code. The prefetch does not help in this case and the overhead really is from the OSSL_PARAMS handling.

**Comment by t8m** [comment]
So yeah, this is indeed a duplicate of #20171

**Comment by mattcaswell** [comment]
> So yeah, this is indeed a duplicate of https://github.com/openssl/openssl/issues/20171

So ....that means it should be fixed in 3.1?

**Comment by t8m** [comment]
> So ....that means it should be fixed in 3.1?

Nope, apparently it isn't. The caching does not seem to help in this particular case.

**Comment by pabuhler** [comment]
With 3.1 if I do not call EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_IVLEN, 12, 0) for each iteration I do see a significant speed up.
So either can call it once during on creation (if it is preserved across new IV's) else initial check the return of EVP_CIPHER_get_iv_length(cipher) and only set it in the loop if it is not 12.
@t8m  do either of these sound ok ?

**Comment by t8m** [comment]
Calling it once is sufficient. It is preserved.

**Comment by t8m** [comment]
We could even test whether there is any change in the value in comparison to the cached one and not invoke the provider API.

**Comment by t8m** [comment]
Can you please test the #20664 patch on top of 3.1?

**Comment by pabuhler** [comment]
@t8m  I have tested and can confirm that this does improve performance but not as much as libsrtp only setting the IV length once. Either way there still seams to be at least a 30% performance penalty (will try to do more comparisons) when encrypting smaller rtp packets, ie audio,  when comparing OpenSSL 1.1 and 3.1 (with patch) . Do you have any other ideas how to improve the performance ? 

@murillo128   I will make a PR on libsrtp with a change to set the IV len only once maybe you can test that.

**Comment by pabuhler** [comment]
@murillo128  can you run your tests again with https://github.com/cisco/libsrtp/pull/649

**Comment by murillo128** [comment]
I am OOO this week, will test it asap next week

**Comment by murillo128** [comment]
@pabuhler I see only a minimal performance improvement using `set-iv-len-once`

```
Node 18
   0 payload size gcm done in 2157292
   0 payload size ctr done in 3311742
 400 payload size gcm done in 2756376
 400 payload size ctr done in 5562947
1200 payload size gcm done in 3654080
1200 payload size ctr done in 10034543

Node 18 set-iv-len-once
   0 payload size gcm done in 1899945
   0 payload size ctr done in 3435551
 400 payload size gcm done in 2537257
 400 payload size ctr done in 5637461
1200 payload size gcm done in 3406418
1200 payload size ctr done in 10074385

```

For reference, this was node 16:

```
Node 16
0    payload size gcm done in 886714
0    payload size ctr done in 2421151
400  payload size gcm done in 1494584
400  payload size ctr done in 4381984
1200 payload size gcm done in 2326866
1200 payload size ctr done in 8568198
``

**Comment by murillo128** [comment]
taking a flamegraph of the `gcm` with 0 payload for checking where the cpu is spent:
![image](https://user-images.githubusercontent.com/1070835/231462635-524245ec-86ef-4147-acc3-12c69a21a9c1.png)

**Comment by t8m** [comment]
As expected the caching and other optimizations can improve the situation a little but it still cannot completely remove the overhead of the OSSL_PARAM handling. Unfortunately there is not much that can be done further apart from completely revamping the OSSL_PARAM concept which is not going to happen any time soon.

We could  probably microoptimize at least some use-cases by reordering the param matching in the ossl_gcm_get_ctx_params() and using the concept of short-circuiting from https://github.com/openssl/openssl/pull/17281

**Comment by pabuhler** [comment]
> Could you please try the draft PR #20721 if it is worth it at all?

I am away from PC this week but can also do follow testing next week. Thanks for trying to help find a solution.

**Comment by murillo128** [comment]
@t8m maybe it is a stupid suggestion, but can't we cache the actual pointer returned by `OSSL_PARAM_locate` and reuse it if the `params` has not been changed to avoid doing the lookup each time?

**Comment by nirbheek** [comment]
I was also wondering whether interned strings are an option, since it would not be an ABI or API break to use them to do the first comparison, and then fallback to doing a string comparison.

**Comment by t8m** [comment]
@murillo128 that could be a good idea.

Using internalized strings could also work.

@pabuhler have you been able to look at #20721 if it helps?

**Comment by pabuhler** [comment]
@t8m  I have tried to test, and I think I see a 10% speedup when encrypting small (128 bytes) amounts of data. Tested both with libsrtp and my basic srtp emulation https://gist.github.com/pabuhler/7beef08da203018ae905944c44c911dd.
But my laptop gives varied results, maybe better to run these tests on a stationary machine but I don't have access just now.
So from my perspective this is a good small step in the right direction but is still ~25% slower than OpenSSL 1.1 .

**Comment by murillo128** [comment]
I have created a potential fix #20812, by  changing the iterator loop in ossl_gcm_get_ctx_params and ossl_gcm_set_ctx_params functions.

On @t8m `speed.c` sample program I have been able to reduce it from `3.5s` to `2.7s`. I have not verified yet what is the performance difference vs openssl 1.1 (@pabuhler  could you check it too?)

**Comment by murillo128** [comment]
umm.. the results are very variable, so would have to double check on monday on a different machine

**Comment by pabuhler** [comment]
@murillo128   I see about the same speed up as the #20721 pr,  does that make sense? Is it possible to combine them ?

**Comment by t8m** [comment]
> @murillo128 I see about the same speed up as the #20721 pr, does that make sense? Is it possible to combine them ?

That's what I would expect. And no, they effectively do the same thing in the benchmark, so they cannot be combined.

**Comment by nhorman** [comment]
marking as inactive, will be closed at the completion of 3.4 dev without further input
